### DISCUSSION TITLE
ModbusTCP Server and Client at the same time on a single ESP32

### DISCUSSION THREAD

**Discussion by Lupo135** [eModbus/eModbus#341] [Q&A] [answered]
I have a running ESP32 module with several tasks running in parallel like web interface, serial metering interface, onewire, littlefs filesystem and internal ADC. 
Is it possible to run ModbusTCP Server and Client at the same time on my a single ESP32?
Do you recommend using sync or async library?

**Comment by Miq1** [comment]
Yes, that is possible. I myself normally would use the regular TCP varieties, I have no profund experience with the async stuff on the ESP32.

**Reply by Lupo135** [reply]
Thank you for the answer and the really great library functions. 
I don't completely understand the difference between sync and async, because both do the work in background and have callbacks when data is received. For my system it is important that there is no blocking task.

**Reply by Miq1** [reply, marked as answer]
There should be no blocking as long as you take care to prevent race conditions between tasks. But please note that the limited resources of an ESP32 in terms of processing power will make real-time Modbus applications somewhat difficult if using many parallel tasks. If you can live with some milliseconds of slack between requests and responses, everything will be fine. If you need to react as fast as possible you may have to play with task priorities etc.
